src/java.base/share/classes/jdk/internal/misc/JitCompiler.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.misc;
+
+import jdk.internal.vm.annotation.IntrinsicCandidate;
+
+/**
+ * Just-in-time-compiler-related queries
+ */
+public class JitCompiler {
+    /**
+     * Determine if {@code expr} can be evaluated to a constant value by the JIT
+     * compiler. For example, {@code isCompileConstant(5)} will be evaluated
+     * to {@code true} by the JIT compiler, while
+     * {@code isCompileConstant(random.nextLong())} will likely be evaluated
+     * to {@code false}.
+     *
+     * <p>Note that the JIT compiler is responsible to change the return value
+     * of this method to {@code true}, the interpreter always returns {@code false}.
+     *
+     * <p>Given the nondeterministic nature of this method, the result of the
+     * program must not depend on the return value of this method. It must be
+     * used as a pure optimization to take advantage of the constant nature of
+     * {@code expr}. E.g. for a runtime variable, looking up a hashmap may be
+     * the most efficient look up method, however, if the look up table is
+     * constant, it may be better to use a chain of if-else in cases where the
+     * input is also a constant.
+     *
+     * @param expr the expression to be evaluated
+     * @return {@code true} if the JIT compiler determines the {@code expr} is
+     *         always evaluated to a constant value, {@code false} otherwise
+     */
+    @IntrinsicCandidate
+    public static boolean isCompileConstant(boolean expr) {
```

Comment:
Here and in other places: probably not `expr`, but just `val` or something?

Reply:
I think of this as an expression that is always evaluated to the same value. The value itself is not interesting, it is the set of values that this expression can take that we are talking about.

Reply:
This seems really weird to me for Java code. The method doesn't get the original "expression" it only gets the value of that expression after it has been evaluated. Is there some kind of weird "magic" happening here?

Reply:
@dholmes-ora Indeed it's a compiler magic, albeit not really weird. While the method execution only receives the evaluated value of `expr`, the method compilation has the expression in its original form. As a result, it can determine the result based on this information.

Reply:
It is still weird to talk about expressions at this level. We really check if the value is constant, like the method name suggests now. Yes, this implicitly tests that the expression that produced that value is fully constant-folded. But that's a detail that we do not need to capture here. Let's rename `expr` -> `val`, and tighten up the javadoc for the method to mention we only test the constness of the final value.

Reply:
I agree. All values are produced by evaluating expressions. In this case we want to query whether a value produced by the compiler evaluating its expression is a constant value (inputs to the expression are constants and the expression had no material side-effects). Meaning if the method returns true then we could use that knowledge in subsequent expressions that may also produce constants or some specific behavior.

Reply:
> the method compilation has the expression in its original form

So the JIT analyses the bytecode used to place the result on the call stack, before the call, and from that determines if the expression were a constant? This kind of self-analysis is not something I was aware of.

Reply:
I see, changed `expr` to `val`.